File: inheritance.rb
class Animal
    attr_accessor :name

    def initialize(name)
        @name = name
    end

    def speak
        return "Hello!"
    end
end

# Dog "is-a" Animal
class Dog < Animal
    attr_reader :breed

    def initialize(name, breed)
        super(name)
        @breed = breed
    end

    def speak
        return "#{super} Woof! I am a #{breed}"
    end
end

class Cat < Animal
end

# Main
fido = Dog.new("Fido", "Border Collie")
felix = Cat.new("Felix")
puts fido.speak
puts felix.speak
# p fido
# p felix

File: inheritance.py
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        return "Hello!"


# Dog "is-a" Animal
class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name)
        self.breed = breed

    def speak(self):
        return f"{super().speak()} Woof! I am a {self.breed}"


class Cat(Animal):
    pass


# Main
fido = Dog("Fido", "Border Collie")
felix = Cat("Felix")
print(fido.speak())
print(felix.speak())
